core/src/main/java/com/google/instrumentation/stats/MutableView.java

```diff
@@ -0,0 +1,155 @@
+import com.google.instrumentation.stats.ViewDescriptor.IntervalViewDescriptor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A mutable version of {@link View}, used for recording stats and start/end time.
+ */
+abstract class MutableView {
+
+  /**
+   * The {@link ViewDescriptor} associated with this {@link View}.
+   */
+  abstract ViewDescriptor getViewDescriptor();
+
+  /**
+   * Applies the given match function to the underlying data type.
+   */
+  abstract <T> T match(
+      Function<MutableDistributionView, T> p0,
+      Function<MutableIntervalView, T> p1);
+
+  /**
+   * Record stats with the given tags.
+   */
+  abstract void record(Map<String, String> tags, double value);
+
+  /**
+   * Convert this {@link MutableView} to {@link View}.
+   */
+  abstract View toView();
+
+  private MutableView() {
+  }
+
+  /**
+   * A {@link MutableView} for recording stats on distribution-based aggregations.
+   */
+  static final class MutableDistributionView extends MutableView {
+
+    /**
+     * Constructs a new {@link MutableDistributionView}.
+     */
+    static MutableDistributionView create(DistributionViewDescriptor distributionViewDescriptor,
+        List<MutableDistribution> distributions, Timestamp start) {
+      return new MutableDistributionView(
+          distributionViewDescriptor, distributions, start);
+    }
+
+    @Override
+    ViewDescriptor getViewDescriptor() {
+      return distributionViewDescriptor;
+    }
+
+    @Override
+    <T> T match(
+        Function<MutableDistributionView, T> p0, Function<MutableIntervalView, T> p1) {
+      return p0.apply(this);
+    }
+
+    @Override
+    void record(Map<String, String> tags, double value) {
+      // TODO(songya): compare tags with tagKeys in ViewDescriptor
+      for (MutableDistribution distribution : mutableDistributions) {
+        distribution.add(value);
+      }
+    }
+
+    @Override
+    final View toView() {
+      final List<DistributionAggregation> distributionAggregations =
+          new ArrayList<DistributionAggregation>();
+      for (MutableDistribution distribution : mutableDistributions) {
+        distributionAggregations.add(
+            DistributionAggregation.create(distribution.getCount(), distribution.getMean(),
+                distribution.getSum(),
+                null, null,  // TODO(songya): convert Range and generate List<Tag>
+                distribution.getBucketCounts()));
+      }
+      return DistributionView.create(distributionViewDescriptor, distributionAggregations, start,
+          Timestamp.fromMillis(System.currentTimeMillis()));
+    }
+
+    /**
+     * Returns start timestamp for this aggregation.
+     */
+    Timestamp getStart() {
+      return start;
+    }
+
+    private final DistributionViewDescriptor distributionViewDescriptor;
+    private final List<MutableDistribution> mutableDistributions;
```

Comment:
Shouldn't mutableDistributions be a map from tag values to MutableDistributions, i.e., `Map<List<TagValue>, Distribution>`?  We will need the tag values when constructing the DistributionAggregations.  Since this PR doesn't use the tag values yet, you could just add a TODO here.